prober/derp.go

```diff
@@ -0,0 +1,379 @@
+
+			for _, to := range region.Nodes {
+				n = fmt.Sprintf("derp/%s/%s/%s/mesh", region.RegionCode, server.Name, to.Name)
+				wantProbes[n] = true
+				if d.probes[n] == nil {
+					log.Printf("adding DERP mesh probe for %s->%s (%s)", server.Name, to.Name, region.RegionName)
+					d.probes[n] = d.p.Run(n, d.meshInterval, labels, d.meshProbeFn(server.HostName, to.HostName))
+				}
+			}
+		}
+	}
+
+	for n, probe := range d.probes {
+		if !wantProbes[n] {
+			log.Printf("removing DERP probe %s", n)
+			probe.Close()
+			delete(d.probes, n)
+		}
+	}
+
+	return nil
+}
+
+func (d *derpProber) probeMesh(from, to string) ProbeFunc {
+	return func(ctx context.Context) error {
+		d.Lock()
+		dm := d.lastDERPMap
+		fromN, ok := d.nodes[from]
+		if !ok {
+			d.Unlock()
+			return fmt.Errorf("could not find derp node %s", from)
+		}
+		toN, ok := d.nodes[to]
+		if !ok {
+			d.Unlock()
+			return fmt.Errorf("could not find derp node %s", to)
+		}
+		d.Unlock()
+
+		// TODO: instead of ignoring latency, export it as a separate metric.
+		_, err := derpProbeNodePair(ctx, dm, fromN, toN)
+		return err
+	}
+}
+
+func (d *derpProber) updateMap(ctx context.Context) error {
+	req, err := http.NewRequestWithContext(ctx, "GET", d.derpMapURL, nil)
+	if err != nil {
+		return nil
+	}
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		d.Lock()
+		defer d.Unlock()
+		if d.lastDERPMap != nil && time.Since(d.lastDERPMapAt) < 10*time.Minute {
+			log.Printf("Error while fetching DERP map, using cached one: %s", err)
+			// Assume that control is restarting and use
+			// the same one for a bit.
+			return nil
+		}
+		return err
+	}
+	defer res.Body.Close()
+	if res.StatusCode != 200 {
+		return fmt.Errorf("fetching %s: %s", d.derpMapURL, res.Status)
+	}
+	dm := new(tailcfg.DERPMap)
+	if err := json.NewDecoder(res.Body).Decode(dm); err != nil {
+		return fmt.Errorf("decoding %s JSON: %v", d.derpMapURL, err)
+	}
+
+	d.Lock()
+	defer d.Unlock()
+	d.lastDERPMap = dm
+	d.lastDERPMapAt = time.Now()
+	d.nodes = make(map[string]*tailcfg.DERPNode)
+	for _, reg := range d.lastDERPMap.Regions {
+		for _, n := range reg.Nodes {
+			if existing, ok := d.nodes[n.HostName]; ok {
+				return fmt.Errorf("derpmap has duplicate nodes: %+v and %+v", existing, n)
+			}
+			d.nodes[n.HostName] = n
+		}
+	}
+	return nil
+}
+
+func (d *derpProber) ProbeUDP(ipaddr string, port int) ProbeFunc {
+	return func(ctx context.Context) error {
+		_, err := derpProbeUDP(ctx, ipaddr, port)
+		return err
+	}
+}
+
+func derpProbeUDP(ctx context.Context, ipStr string, port int) (latency time.Duration, err error) {
+	pc, err := net.ListenPacket("udp", ":0")
+	if err != nil {
+		return 0, err
+	}
+	defer pc.Close()
+	uc := pc.(*net.UDPConn)
+
+	tx := stun.NewTxID()
+	req := stun.Request(tx)
+
+	if port == 0 {
+		port = 3478
+	}
+	for {
+		ip := net.ParseIP(ipStr)
+		_, err := uc.WriteToUDP(req, &net.UDPAddr{IP: ip, Port: port})
+		if err != nil {
+			return 0, err
+		}
+		buf := make([]byte, 1500)
```

Comment:
I imagine this is 1500 to be the largest a packet can be, but 1500 isn't the largest a packet can be.

If a DERP is running in AWS it might use jumbo frames to get 9001 bytes.

Since these are UDP frames, I think they could be up to 64K and just get fragmented. I don't see anything which would prevent reassembled fragmented frames from being received.

All of which is to say that I imagine making the buffer as large as we expect a STUN frame to be is probably sufficient, and if we want to handle a maximum size frame then I'm not sure 1500 is enough.

Reply:
I looked at the [original commit](https://github.com/tailscale/tailscale/commit/13ef8e3c06de619d25eea3611027d2b1143cc9cf) and it's unclear why 1500 was chosen, but according to [STUN RFC](https://www.rfc-editor.org/rfc/rfc5389#section-7.1) the UDP messages should not exceed path MTU, so I guess 1500 might have been a reasonable upper bound in a non-jumbo frame situation.

In practice we are sending a Binding request that is around 40 bytes in size, with the expected binding response being similarly small. Using a 64K buffer for the response seems a bit wasteful, but I have increased this to 9000 to accommodate jumbo frames and added a comment explaining rationale.